### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by hamidazim321**
# PR summary
We have written unit tests for all methods in:
- Book
- Person
- Student
- Teacher
- Classroom
- Rental
- all classes that we created while implementing the Decorator design pattern 

## Changes made after previous review
- I have added tests for Decorator and Nameable classes
- I have updated tests for Person class to include all methods
- Since ```of_age?``` method in Person class is a private method and is used in ```can_use_services?```, I have added test for  ```can_use_services?``` method to check the ```of_age?``` class is working correctly
- I have updated the tests for Book class to test the add_rental method
- I have updated the tests for Student class to test the play_hooky method
- Overall, I have made all the required changes required by the previous reviewer

### CODE REVIEW

File: spec/person_spec.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    expect(person.age).to eq 18
+    expect(person.name).to eq 'unknown'
+    expect(person.parent_permission).to eq true
+  end
+
+  it 'should get correct name' do
+    person = Person.new(25, 'John Doe')
+
+    expect(person.correct_name).to eq 'John Doe'
+  end
```
Comment (OyePriscilla) [inline spec/person_spec.rb:15]: Awesome implementation for the person's class, kindly add test for the methods highlighted below as well as stated in the project requirement. ℹ️ 

CAROL

File: spec/book_spec.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+require_relative '../book'
+
+describe 'testing the book class' do
+  it 'should initialize with title and author' do
+    book = Book.new('The Great Gatsby', 'F. Scott Fitzgerald')
+    expect(book.title).to eql 'The Great Gatsby'
+    expect(book.author).to eql 'F. Scott Fitzgerald'
+  end
```
Comment (OyePriscilla) [inline spec/book_spec.rb:8]: - [ ] Please add test for the `add_rental` method in book.rb file. 👍

File: spec/student_spec.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+require_relative '../student'
+
+describe 'testing student class' do
+  it 'should generate both getters and setters' do
+    student = Student.new(18)
+    student.age = 20
```
Comment (OyePriscilla) [inline spec/student_spec.rb:6]: - [ ] Kindly add test to check when initializing the `Student` class (test for constructor method) as you rightly did for book_spec.rb and person_spec.rb files. 👍

- [ ] The test to check if students can `play_hooky` is missing as well. Kindly implement it. 👍

File: spec/teacher_spec.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+require_relative '../teacher'
+describe Teacher do
+  context 'When testing the Teacher class' do
+    it 'Should always return true on can_use_services? method' do
+      teacher = Teacher.new(22, 'CS', 'John')
+      expect(teacher.can_use_services?).to be true
+    end
+  end
```
Comment (OyePriscilla) [inline spec/teacher_spec.rb:8]: - [ ] Kindly add test to check when initializing the `Teacher` class (test for constructor method) as you rightly did for `book_spec.rb` and `person_spec.rb` files. 👍

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @hamidazim321 and @Fombi-Favour 👏 👏 👏 👏, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

so-close

## Highlights
Write unit tests for an existing codebase using Rspec. ✔️ 

## Required Changes ♻️

- [ ] Please kindly implement tests for the methods in `nameable.rb and decorator.rb` files as stated in the requirement. 👍 

CAROL

_Check the comments under the review also._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by skyv26 — APPROVED** [review]
## Hi @hamidazim321 and @Fombi-Favour, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## STATUS: APPROVED 🟢 



## Highlights
- Implement the `tests` for the method as requested by previous code reviewer ✅ 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

## ⚠️  WARNING ⚠️ 
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**